server/src/main/java/org/apache/druid/discovery/BrokerClient.java

```diff
@@ -0,0 +1,143 @@
+import com.google.inject.Inject;
+import org.apache.druid.guice.annotations.EscalatedGlobal;
+import org.apache.druid.java.util.common.IOE;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.java.util.common.RE;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.java.util.http.client.HttpClient;
+import org.apache.druid.java.util.http.client.Request;
+import org.apache.druid.java.util.http.client.response.StringFullResponseHandler;
+import org.apache.druid.java.util.http.client.response.StringFullResponseHolder;
+import org.jboss.netty.channel.ChannelException;
+import org.jboss.netty.handler.codec.http.HttpMethod;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * This class facilitates interaction with Broker.
+ */
+public class BrokerClient
+{
+  private final Logger log = new Logger(BrokerClient.class);
+  private static final int MAX_RETRIES = 5;
+
+  private final HttpClient httpClient;
+  private final DruidNodeDiscovery druidNodeDiscovery;
+
+  @Inject
+  public BrokerClient(
+      @EscalatedGlobal HttpClient httpClient,
+      DruidNodeDiscoveryProvider druidNodeDiscoveryProvider
+  )
+  {
+    this.httpClient = httpClient;
+    this.druidNodeDiscovery = druidNodeDiscoveryProvider.getForNodeRole(NodeRole.BROKER);
+  }
+
+  /**
+   * Creates and returns a {@link Request} after choosing a broker.
+   */
+  public Request makeRequest(HttpMethod httpMethod, String urlPath) throws IOException
+  {
+    String host = ClientUtils.pickOneHost(druidNodeDiscovery);
+
+    if (host == null) {
+      throw new IOE("No known server.");
+    }
+    return new Request(httpMethod, new URL(StringUtils.format("%s%s", host, urlPath)));
+  }
+
+  public String sendQuery(Request request) throws Exception
+  {
+    StringFullResponseHandler responseHandler = new StringFullResponseHandler(StandardCharsets.UTF_8);
+
+    for (int counter = 0; counter < MAX_RETRIES; counter++) {
+      final StringFullResponseHolder fullResponseHolder;
+
+      try {
+        try {
+          fullResponseHolder = httpClient.go(request, responseHandler).get();
+        }
+        catch (ExecutionException e) {
+          // Unwrap IOExceptions and ChannelExceptions, re-throw others
+          Throwables.propagateIfInstanceOf(e.getCause(), IOException.class);
+          Throwables.propagateIfInstanceOf(e.getCause(), ChannelException.class);
+          throw new RE(e, "HTTP request to [%s] failed", request.getUrl());
+        }
+      }
+      catch (IOException | ChannelException ex) {
+        // can happen if the node is stopped.
+        log.warn(ex, "Request [%s] failed.", request.getUrl());
```

Comment:
This would log it after each retry. Since the retries are happening in a short span, there's a high likelihood that we would be posting the same stack over and over. This should log once after all the retries are exhausted. If you refactor it to RetryUtils, I think it also handles that for you.